From e628ee560a6cf7e7bd5011a6fda615afc3fec292 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 11 Mar 2020 10:27:48 +0800
Subject: [PATCH 36/81] modesetting: Add option NoEDID

Some monitors might have broken EDID, add an option to ignore that.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 hw/xfree86/drivers/modesetting/driver.c          | 1 +
 hw/xfree86/drivers/modesetting/driver.h          | 1 +
 hw/xfree86/drivers/modesetting/drmmode_display.c | 3 ++-
 3 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/hw/xfree86/drivers/modesetting/driver.c b/hw/xfree86/drivers/modesetting/driver.c
index 1501e5043..08ae861c8 100644
--- a/hw/xfree86/drivers/modesetting/driver.c
+++ b/hw/xfree86/drivers/modesetting/driver.c
@@ -159,6 +159,7 @@ static const OptionInfoRec Options[] = {
     {OPTION_FLIP_FB, "FlipFB", OPTV_STRING, {0}, FALSE},
     {OPTION_FLIP_FB_RATE, "MaxFlipRate", OPTV_INTEGER, {0}, 0},
     {OPTION_BIND_CURRENT, "BindCurrent", OPTV_BOOLEAN, {0}, FALSE},
+    {OPTION_NO_EDID, "NoEDID", OPTV_BOOLEAN, {0}, FALSE},
     {-1, NULL, OPTV_NONE, {0}, FALSE}
 };
 
diff --git a/hw/xfree86/drivers/modesetting/driver.h b/hw/xfree86/drivers/modesetting/driver.h
index 23864ec74..44f7c84b9 100644
--- a/hw/xfree86/drivers/modesetting/driver.h
+++ b/hw/xfree86/drivers/modesetting/driver.h
@@ -65,6 +65,7 @@ typedef enum {
     OPTION_FLIP_FB,
     OPTION_FLIP_FB_RATE,
     OPTION_BIND_CURRENT,
+    OPTION_NO_EDID,
 } modesettingOpts;
 
 typedef struct
diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index eb1258d45..06337f196 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -2793,7 +2793,8 @@ drmmode_output_get_modes(xf86OutputPtr output)
     drmmode_output->edid_blob =
         koutput_get_prop_blob(drmmode->fd, koutput, "EDID");
 
-    if (drmmode_output->edid_blob) {
+    if (!xf86ReturnOptValBool(drmmode->Options, OPTION_NO_EDID, FALSE) &&
+        drmmode_output->edid_blob) {
         mon = xf86InterpretEDID(output->scrn->scrnIndex,
                                 drmmode_output->edid_blob->data);
         if (mon && drmmode_output->edid_blob->length > 128)
-- 
2.38.5

